woocommerce/woocommerce-gateway-stripe
# Fix legacy method names in order meta

Fixes #2932 

## Changes proposed in this Pull Request:
- In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2934/commits/3fbd3d524d03ac9d3464c3ee858333a2f4ffe98e I have reverted the subscription check for Multibanco. I had added this check in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2870 to hide Multibanco as it was showing up as a payment method while purchasing a subscription when UPE was enabled. However, upon testing in `trunk` I found that trunk has the same behavior for Multibanco in UPE and all other methods when UPE is disabled. In short, the existing behavior in the latest releases are, when you have a subscription product in cart, the non-UPE methods using Sources API is available as an option to checkout but the subscription will be created as a manually renewed subscription.
- In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2934/commits/25594a8c3187ebf70a6b48a9d75e4a9f359db52b reverted the changes I made in PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2839. I removed all the stripe methods from `add_gateways` [method](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/woocommerce-gateway-stripe.php#L242) which was called by `woocommerce_payment_gateways` so that they don't appear on the **WooCommerce > Settings > Payments** page. As a result `WC()->payment_gateways->payment_gateways()` does not have these Stripe methods listed. As a result, [here](https://github.com/woocommerce/woocommerce/blob/8c55e53772a20a84f7071089dbf1675bc7dceb02/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php#L172) the order [can not find the stripe payment method object](https://github.com/woocommerce/woocommerce/blob/8c55e53772a20a84f7071089dbf1675bc7dceb02/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php#L177) (just has the id) and finally just renders the id [here](https://github.com/woocommerce/woocommerce/blob/8c55e53772a20a84f7071089dbf1675bc7dceb02/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php#L210). This is the reason behind the order meta payment method name issue in #2932. (There could be other places implicated by `WC()->payment_gateways->payment_gateways()` not having the stripe payment methods, though I have not found others yet)
- As in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2934/commits/25594a8c3187ebf70a6b48a9d75e4a9f359db52b I am adding the stripe payment methods to the list again, they are visible on the **WooCommerce > Settings > Payments** page. So in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2934/commits/803835fa6b2839307328e8e846192a3dfad3c3c0 I am hiding them using css.
- To show the payment methods in the order saved inside Stripe settings page and not in the order of the outer settings page (where they are hidden), in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2934/commits/baf0589ae51b169e467b9d0fab5283aaaf7b83f7 I am unsettling them from the gateway object before using them to the expected position.


## Testing instructions
#### Reproduce the bug
- Go to the `develop` branch.
- Disable the `New checkout experience` from the Stripe settings page.
- Set Euro as store currency and enable a few payment methods on the settings page (like EPS, giropay etc.)
- Click the customize button, set a title for the enabled methods or keep it as before.
- As a shopper, add a product to the cart and checkout with any Stripe enabled methods except `card`.
- After placing the order, in your wp dashboard go to this order edit page. 
- Notice that the payment method name is displayed as `stripe_xyz` instead of the method title.

![wpm](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/68ca203c-5964-462a-8079-ad48bca09284)

#### Test the fix
- Go to the `fix/2932-non-upe-method-in-order-and-checkout` branch (this branch).
- Disable the `New checkout experience` from the Stripe settings page.
- Set Euro as store currency and enable a few payment methods on the settings page (like EPS, giropay etc.)
- Click the customize button, set a title for the enabled methods or keep it as before.
- As a shopper, add a product to the cart and checkout with any Stripe enabled methods except `card`.
- After placing the order, in your wp dashboard go to this order edit page. 
- Make sure that the payment method name is displayed as the one you have set as the title.

![cpm](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/b150cc63-17ba-4f7f-b630-4359fa603782)

#### Payment method order in checkout page
- Go to the `fix/2932-non-upe-method-in-order-and-checkout` branch (this branch).
- Disable the `New checkout experience` from the Stripe settings page.
- Reorder payment methods and customize some of their names.
- Reorder the main Stripe method in **WooCommerce > Settings > Payments**. (cod, stripe, cheque)
- As a shopper, go to the checkout page and ensure that the ordering is correct. Expected ordering would be cod, all the Stripe methods as ordered inside the Stripe settings page, cheque.
- Reorder payment methods in Stripe settings page, refresh the checkout page as a shopper and make sure the order is correct.

#### Add payment method page
- Enable UPE
- Enable a few methods including Card, Sepa, iDeal and Bancontact.
- Go to **My Account > Add payment method** page. 
- When you try to add a payment method Card, Sepa, iDeal and Bancontact should be available as options.
- Disable UPE
- Enable a few methods including Card and Sepa
- Go to **My Account > Add payment method** page. 
- When you try to add a payment method Card and Sepa.
